File: Flyweight/Go/flyweight/FlyweightFactory.go
package flyweight

type FlyweightFactory struct {
  fmMap map[string]*AuthorizationFactory
}

var instance *FlyweightFactory

func GetInstance() *FlyweightFactory {
  if instance == nil {
    instance = &FlyweightFactory{}
  }

  return instance
}

func (this *FlyweightFactory) GetFlyweight(key string) *AuthorizationFactory {
  if this.fmMap == nil {
    this.fmMap = make(map[string]*AuthorizationFactory)
  }

  if _, ok := this.fmMap[key]; !ok {
    this.fmMap[key] = NewAuthorizationFactory(key)
  }

  return this.fmMap[key]
}

File: Flyweight/PHP/FlyweightFactory.php
<?php
  class FlyweightFactory {
    private static $instance;

    private $fwMap;

    private function __construct() {
      $this->fwMap = array();
    }

    public static function getInstance() {
      if (self::$instance === null) {
        self::$instance = new self();
      }

      return self::$instance;
    }

    public function getFlyweight($key) {
      if (!isset(self::$instance->fwMap[$key])) {
        self::$instance->fwMap[$key] = new AuthorizationFlyweight($key);
      }

      return self::$instance->fwMap[$key];
    }

  }
